File: src.rs/src/algorithms.rs
use std::collections::HashMap;
use rand::{rngs::StdRng, Rng, SeedableRng};

use crate::exchange::{
    cancel_order, get_mid_price, submit_order, LimitOrderBook, MarketDataEvent, Order, OrderType,
    Side,
};

pub struct MarketDataBus {
    subscribers: Vec<Box<dyn FnMut(&MarketDataEvent)>>,
}

impl MarketDataBus {
    pub fn new() -> Self {
        Self { subscribers: Vec::new() }
    }

    pub fn subscribe<F>(&mut self, handler: F)
    where
        F: FnMut(&MarketDataEvent) + 'static,
    {
        self.subscribers.push(Box::new(handler));
    }

    pub fn emit(&mut self, event: &MarketDataEvent) {
        for handler in self.subscribers.iter_mut() {
            handler(event);
        }
    }
}

#[derive(Default)]
pub struct AlgoState {
    pub inventory: i64,
    pub cash: f64,
    pub trades: i64,
}

impl AlgoState {
    pub fn pnl(&self) -> f64 {
        self.cash
    }
}

pub struct AlgorithmBase {
    pub algo_id: i64,
    pub state: AlgoState,
    orders: HashMap<i64, Order>,
    next_order_id: i64,
}

impl AlgorithmBase {
    pub fn new(algo_id: i64) -> Self {
        Self {
            algo_id,
            state: AlgoState::default(),
            orders: HashMap::new(),
            next_order_id: algo_id * 1_000_000,
        }
    }

    pub fn next_order_id(&mut self) -> i64 {
        self.next_order_id += 1;
        self.next_order_id
    }

    pub fn on_market_data(&mut self, event: &MarketDataEvent) {
        match event {
            MarketDataEvent::OrderExecuted { order_id, exec_quantity, exec_price, .. } => {
                if let Some(order) = self.orders.get(order_id) {
                    match order.side {
                        Side::Buy => {
                            self.state.inventory += exec_quantity;
                            self.state.cash -= *exec_quantity as f64 * exec_price;
                        }
                        Side::Sell => {
                            self.state.inventory -= exec_quantity;
                            self.state.cash += *exec_quantity as f64 * exec_price;
                        }
                    }
                    self.state.trades += 1;
                }
            }
            _ => {}
        }
    }

    pub fn send_order(&mut self, book: &mut LimitOrderBook, order: Order, current_time: f64) {
        self.orders.insert(order.id, order.clone());
        submit_order(book, order, current_time);
    }

    pub fn cancel_order(&mut self, book: &mut LimitOrderBook, order_id: i64, current_time: f64) {
        cancel_order(book, order_id, current_time);
    }
}

pub struct RandomTrader {
    base: AlgorithmBase,
    rng: StdRng,
    max_qty: i64,
    price_band: f64,
    market_prob: f64,
}

impl RandomTrader {
    pub fn new(algo_id: i64, seed: u64) -> Self {
        Self {
            base: AlgorithmBase::new(algo_id),
            rng: StdRng::seed_from_u64(seed),
            max_qty: 5,
            price_band: 1.0,
            market_prob: 0.3,
        }
    }

    pub fn step(&mut self, book: &mut LimitOrderBook, current_time: f64) {
        let side = if self.rng.gen::<f64>() < 0.5 { Side::Buy } else { Side::Sell };
        let qty = self.rng.gen_range(1..=self.max_qty);
        let order_type = if self.rng.gen::<f64>() < self.market_prob {
            OrderType::Market
        } else {
            OrderType::Limit
        };
        let mut price = f64::NAN;
        if order_type == OrderType::Limit {
            let mid = get_mid_price(book);
            let offset = (self.rng.gen::<f64>() - 0.5) * 2.0 * self.price_band;
            price = (mid + offset).max(book.tick_size);
        }

        let order = Order {
            id: self.base.next_order_id(),
            trader_id: self.base.algo_id,
            side,
            order_type,
            price,
            quantity: qty,
            timestamp: current_time,
            filled: 0,
            iceberg_display_qty: None,
            iceberg_hidden_qty: 0,
            stop_price: None,
            time_in_force: "gtc".to_string(),
            post_only: false,
        };
        self.base.send_order(book, order, current_time);
    }
}

pub struct MarketMaker {
    base: AlgorithmBase,
    spread: f64,
    size: i64,
    refresh_interval: f64,
    bid_id: Option<i64>,
    ask_id: Option<i64>,
}

impl MarketMaker {
    pub fn new(algo_id: i64) -> Self {
        Self {
            base: AlgorithmBase::new(algo_id),
            spread: 0.02,
            size: 5,
            refresh_interval: 0.2,
            bid_id: None,
            ask_id: None,
        }
    }

    pub fn step(&mut self, book: &mut LimitOrderBook, current_time: f64) {
        if let Some(id) = self.bid_id {
            self.base.cancel_order(book, id, current_time);
        }
        if let Some(id) = self.ask_id {
            self.base.cancel_order(book, id, current_time);
        }

        let mid = get_mid_price(book);
        let bid_price = (mid - self.spread / 2.0).max(book.tick_size);
        let ask_price = (mid + self.spread / 2.0).max(book.tick_size);

        let bid = Order {
            id: self.base.next_order_id(),
            trader_id: self.base.algo_id,
            side: Side::Buy,
            order_type: OrderType::Limit,
            price: bid_price,
            quantity: self.size,
            timestamp: current_time,
            filled: 0,
            iceberg_display_qty: None,
            iceberg_hidden_qty: 0,
            stop_price: None,
            time_in_force: "gtc".to_string(),
            post_only: true,
        };
        let ask = Order {
            id: self.base.next_order_id(),
            trader_id: self.base.algo_id,
            side: Side::Sell,
            order_type: OrderType::Limit,
            price: ask_price,
            quantity: self.size,
            timestamp: current_time,
            filled: 0,
            iceberg_display_qty: None,
            iceberg_hidden_qty: 0,
            stop_price: None,
            time_in_force: "gtc".to_string(),
            post_only: true,
        };

        self.bid_id = Some(bid.id);
        self.ask_id = Some(ask.id);
        self.base.send_order(book, bid, current_time);
        self.base.send_order(book, ask, current_time);
    }
}

File: src/algorithms.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Dict, List, Optional
import math
import random

from src.exchange import (
    BookUpdateEvent,
    Order,
    OrderExecutedEvent,
    TradeEvent,
    cancel_order,
    get_mid_price,
    submit_order,
)
from src.sim import Simulation

# interface fortrading algorithms that interact with the exchange

class MarketDataBus:
    # simple pub/sub for market events
    def __init__(self) -> None:
        self._subscribers: List[Callable] = []

    def subscribe(self, handler: Callable) -> None:
        self._subscribers.append(handler)

    def __call__(self, book, event, current_time) -> None:
        # broadcast event to all subscribers
        for handler in self._subscribers:
            handler(event)


@dataclass
class AlgoState:
    # tracks what the algo owns and how it's doing
    inventory: int = 0  # how many shares we have
    cash: float = 0.0  # money in/out from trades
    trades: int = 0  # number of trades executed

    @property
    def pnl(self) -> float:
        # profit/loss is just our cash (ignoring inventory value)
        return self.cash


class Algorithm:
    # base class for trading algorithms
    def __init__(self, algo_id: int, book, sim: Simulation) -> None:
        self.algo_id = algo_id
        self.book = book
        self.sim = sim
        self.state = AlgoState()
        self._orders: Dict[int, Order] = {}
        self._next_order_id = algo_id * 1_000_000  # unique id range per algo

    def next_order_id(self) -> int:
        # generate unique order ids
        self._next_order_id += 1
        return self._next_order_id

    def on_market_data(self, event) -> None:
        # handle market events and update our state
        if isinstance(event, OrderExecutedEvent):
            order = self._orders.get(event.order_id)
            if order is None:
                return
            # update inventory and cash based on trade
            if order.side == "buy":
                self.state.inventory += event.exec_quantity
                self.state.cash -= event.exec_quantity * event.exec_price
            else:
                self.state.inventory -= event.exec_quantity
                self.state.cash += event.exec_quantity * event.exec_price
            self.state.trades += 1
        if isinstance(event, TradeEvent):
            return

    def send_order(self, order: Order, current_time: float) -> None:
        # send order to the exchange
        self._orders[order.id] = order
        submit_order(self.book, order, current_time)

    def cancel_order(self, order_id: int, current_time: float) -> None:
        # cancel an order
        cancel_order(self.book, order_id, current_time)


class RandomTrader(Algorithm):
    # sends random orders at regular intervals
    def __init__(
        self,
        algo_id: int,
        book,
        sim: Simulation,
        seed: int = 7,
        interval: float = 0.1,  # time between orders
        max_qty: int = 5,
        price_band: float = 1.0,  # how far from mid to place limits
        market_prob: float = 0.3,  # chance of market order
    ) -> None:
        super().__init__(algo_id, book, sim)
        self.interval = interval
        self.max_qty = max_qty
        self.price_band = price_band
        self.market_prob = market_prob
        self.rng = random.Random(seed)
        self.sim.schedule_event(self.sim.current_time, self._step)

    def _step(self) -> None:
        # create and send a random order
        side = "buy" if self.rng.random() < 0.5 else "sell"
        qty = self.rng.randint(1, self.max_qty)
        order_type = "market" if self.rng.random() < self.market_prob else "limit"
        price = math.nan
        if order_type == "limit":
            # pick price near mid with some random offset
            mid = get_mid_price(self.book)
            offset = (self.rng.random() - 0.5) * 2 * self.price_band
            price = max(self.book.tick_size, mid + offset)

        order = Order(
            id=self.next_order_id(),
            trader_id=self.algo_id,
            side=side,
            order_type=order_type,
            price=price,
            quantity=qty,
            timestamp=self.sim.current_time,
        )
        self.send_order(order, self.sim.current_time)
        # schedule next order
        self.sim.schedule_event(self.sim.current_time + self.interval, self._step)


class MarketMaker(Algorithm):
    # posts bid/ask quotes on both sides of the book
    def __init__(
        self,
        algo_id: int,
        book,
        sim: Simulation,
        spread: float = 0.02,  # distance between our bid and ask
        size: int = 5,  # qty on each side
        refresh_interval: float = 0.2,  # how often to update quotes
    ) -> None:
        super().__init__(algo_id, book, sim)
        self.spread = spread
        self.size = size
        self.refresh_interval = refresh_interval
        self.bid_id: Optional[int] = None
        self.ask_id: Optional[int] = None
        self.sim.schedule_event(self.sim.current_time, self._refresh)

    def on_market_data(self, event) -> None:
        super().on_market_data(event)
        if isinstance(event, BookUpdateEvent):
            return

    def _refresh(self) -> None:
        # cancel old quotes and post new ones
        if self.bid_id is not None:
            self.cancel_order(self.bid_id, self.sim.current_time)
        if self.ask_id is not None:
            self.cancel_order(self.ask_id, self.sim.current_time)

        # calculate new bid/ask prices around mid
        mid = get_mid_price(self.book)
        bid_price = max(self.book.tick_size, mid - self.spread / 2)
        ask_price = max(self.book.tick_size, mid + self.spread / 2)

        # create bid and ask orders
        bid = Order(
            id=self.next_order_id(),
            trader_id=self.algo_id,
            side="buy",
            order_type="limit",
            price=bid_price,
            quantity=self.size,
            timestamp=self.sim.current_time,
            post_only=True,  # won't cross the spread
        )
        ask = Order(
            id=self.next_order_id(),
            trader_id=self.algo_id,
            side="sell",
            order_type="limit",
            price=ask_price,
            quantity=self.size,
            timestamp=self.sim.current_time,
            post_only=True,
        )

        self.bid_id = bid.id
        self.ask_id = ask.id
        self.send_order(bid, self.sim.current_time)
        self.send_order(ask, self.sim.current_time)
        # schedule next refresh
        self.sim.schedule_event(self.sim.current_time + self.refresh_interval, self._refresh)
